### PR TITLE
fix to make work with php8.1

### PR DESCRIPTION
**Pull request by cbtrident**
this work but some can be improve more like the date/time for files
but it is it is importing and it just the need to look better but am not sure how to improve that 
but it working

### CODE REVIEW

**Review by hazzy — COMMENTED** [review]
The getLastSaved function is called twice in the Magmi code that I could find:- 

\magmi\web\magmi_config_setup.php(**151**): Saved:<?php echo $conf->getLastSaved("%c")
magmi\web\magmi_profile_panel.php(**32**): Saved:<?php echo $eplconf->getLastSaved("%c")

It uses “%c” as the desired format. For strfmttime “%c”  passed means “Preferred date and time stamp based on locale”

Example: Tue Feb 5 00:45:10 2009 for February 5, 2009 at 12:45:10 AM

For the PHP date function “%c” is an invalid format – see  https://www.php.net/manual/en/datetime.format.php  for valid values

I suggest you edit the two lines in the files highlighted above and change “%c” to “r” producing a  » RFC 2822/» RFC 5322 formatted date
Example: Thu, 21 Dec 2000 16:01:07 +0200

I don’t use the front end myself so I am unable to test. 

Thanks for the other changes! Most helpful. 

John

**Comment by hazzy** [comment]
If it helps anyone else, I found another issue with engines/magmi_productimportengine.php runing under php 8.1

It issues a warning :- 

PHP Warning:  Trying to access array offset on value of type int in ..../magmi/engines/magmi_productimportengine.php on line 1869

My code change to fix the issue is to add a test to see if $item["qty"] is actually an array

>                // if magmi_qty_absolute flag is not set, then use standard "relative" qty parsing.
>                 if (!isset($item["magmi_qty_absolute"]) || $item["magmi_qty_absolute"] == 0) {
>                     // test for relative qty
>                     //  PHP 8.1 change - test for array - we get an error is $item["qty"] is not an array 
>                     if (is_array($item["qty"])) {
>                       if ($item["qty"][0] == "+" || $item["qty"][0] == "-") {
>                          $relqty = getRelative($item["qty"]);
>                       }
>                     }

**Comment by cbtrident** [comment]
thanks for that Hazzy will add your change to my branch
will test the date change later as am not using php8.1 right now had other problem 
but think now they fixed i could use but anyway i will give this branch alive till the main guys are back

**Comment by Nofear262** [comment]
Just tested it, and it seems to work. However the import process does not generate a process bar. It is just a static page.
I can only cancel, wich does nothing or go back to previous page, but the process seem to do it's work in the back-end.

On the magmi page i have this error:
Deprecated: Optional parameter $dorewrite declared before required parameter $store is implicitly treated as a required parameter in /magento/magmi/plugins/extra/itemprocessors/itemindexer/otfindexer.php on line 232

thanks for making it work in php 8.1 !

**Review by hazzy — COMMENTED** [review]
Thanks cbtrident. 

If you want to, can you fix the comment text. I meant "if" and not "is"

   //  PHP 8.1 change - test for array - we get an error if $item["qty"] is not an array

**Comment by cbtrident** [comment]
lol that will cost you :-P

**Comment by hazzy** [comment]
All for the good of the public!

**Comment by cbtrident** [comment]
anyway thanks,   be nice   if the main people do the fix but least we got something that work still other error to fix but right now dont have much time to cehck

**Comment by hazzy** [comment]
@Nofear262 I had a look at your error. See https://php.watch/versions/8.0/deprecate-required-param-after-optional 

I could not find anywhere in the Magmi code that the builProductUrlRewrite function is called but you might like to check. 

If that is the case then you can fix the issue by changing the order of the parameters in the function 

viz on line 232: 

` public function builProductUrlRewrite($pid, $store, $dorewrite=false )`

If it is called, then the calling scripts will need to be changed of course to reflect the modified parameter order

John 

scratch that! - it is called on line 347 of the otfindexer.php script so that line needs to be changed to reflect the modified parameter order thus

        $purlk = $this->builProductUrlRewrite($pid, $store, $dorewrite );

File: magmi/inc/magmi_config.php [hunk 1]
```diff
@@ -37,7 +37,8 @@ public function getConfFile()
     public function getLastSaved($fmt)
```
Comment (tmotyl) [inline magmi/inc/magmi_config.php]: what is the type of $fmt? can you add a doc block while on it?

File: magmi/inc/magmi_config.php [hunk 2]
```diff
@@ -37,7 +37,8 @@ public function getConfFile()
...
     {
         if (file_exists($this->inifile)) {
-            $lastsaved = strftime($fmt, filemtime($this->inifile));
+           /* $lastsaved = strftime($fmt, filemtime($this->inifile));*/
```
Comment (tmotyl) [inline magmi/inc/magmi_config.php]: remove the comment

File: magmi/inc/magmi_config.php [hunk 3]
```diff
@@ -37,7 +37,8 @@ public function getConfFile()
...
+	$lastsaved = date($fmt, filemtime($this->inifile));
```
Comment (tmotyl) [inline magmi/inc/magmi_config.php]: wrong indentation

File: magmi/plugins/5b5/general/attributesetimport/multi_dim_array.php [hunk 4]
```diff
@@ -146,6 +147,7 @@ public function offsetExistsPartly($name, $partlyOk = true)
     /**
      * @param string $name
      */
+#[\ReturnTypeWillChange]
```
Comment (tmotyl) [inline magmi/plugins/5b5/general/attributesetimport/multi_dim_array.php:150]: indentaion

File: magmi/engines/magmi_productimportengine.php [hunk 5]
```diff
@@ -1864,10 +1864,12 @@ public function updateStock($pid, $item, $isnew)
                 // if magmi_qty_absolute flag is not set, then use standard "relative" qty parsing.
                 if (!isset($item["magmi_qty_absolute"]) || $item["magmi_qty_absolute"] == 0) {
                     // test for relative qty
-                    if ($item["qty"][0] == "+" || $item["qty"][0] == "-") {
-                        $relqty = getRelative($item["qty"]);
-                    }
-                }
+                    //  PHP 8.1 change - test for array - we get an error if $item["qty"] is not an array 
+                	if (is_array($item["qty"])) {
+                  	if ($item["qty"][0] == "+" || $item["qty"][0] == "-") {
```
Comment (tmotyl) [inline magmi/engines/magmi_productimportengine.php:1869]: indentation

**Review by tmotyl — CHANGES_REQUESTED** [review]

**Comment by cbtrident** [comment]
ok done the small changes hope it right

File: magmi/inc/magmi_config.php (hunk 1 above)
Comment (cbtrident) [reply magmi/inc/magmi_config.php]: mmm this is not my code so not sure what you want ,think who did that code should do that

Comment (tmotyl) [reply magmi/inc/magmi_config.php]: I know that its not your code, but you're changing line 40 `$lastsaved = date($fmt, filemtime($this->inifile));`
so you must know what type $fmt is, right?

**Comment by Nofear262** [comment]
> 

Thanks Hazzy, 
I indeed found out that chaninging the order of params solved the issue. But i am not a coder at all.
So i will definately change line 347 also.
Thank you very much.

Also the process page started working yesterday out of the blue.
We imported +2500 prices and stock, very happy with that.

I looked for days on Github to find this thread, so i hope the original magmi 2.x version gets updated !

File: magmi/inc/magmi_config.php (hunk 1 above)
Comment (cbtrident) [reply magmi/inc/magmi_config.php:37]: you mean strtime is not right so changed to date but the part inside are the same
but it from a list on php

Comment (cbtrident) [reply magmi/inc/magmi_config.php]: sorry it was strftime is been drop so it chage to date but the format are the same but i see what $fmt is doing 
as it could be wrong but it what was used to setup the strftime format
so it could be Y-m-d in date
am stilling learning php 8

Comment (hazzy) [reply magmi/inc/magmi_config.php]: I have added the doc block as per tmotyl's request 

Also set a default for the string format to "r"

replacement code for the whole function  below - cbtrident can you adopt it please? 

```
/**
 * Get the file last saved date and time of a file
 *
 * @param string $fmt - any of the PHP time format strings documented here
 * https://www.php.net/manual/en/datetime.format.php 
 * 
 * A sensible one to use would be "r" producing a » RFC 2822/» RFC 5322 formatted date
 * Example of the returned sting: Thu, 21 Dec 2000 16:01:07 +0200
 *
 * @return string
 */    
    
    public function getLastSaved($fmt = "r")
    {
        if (file_exists($this->inifile)) {
            $lastsaved=date($fmt, filemtime($this->inifile));
        } else {
            $lastsaved="never";
        }
        return $lastsaved;
    }


```

Comment (cbtrident) [reply magmi/inc/magmi_config.php]: ok have add

**Comment by onedolphin** [comment]
Hey, I'm not a coder, where can I get this update and try it on php 8.1? Thanks to all

**Comment by cbtrident** [comment]
> Hey, I'm not a coder, where can I get this update and try it on php 8.1? Thanks to all

https://github.com/cbtrident/magmi-m2 is my site for this but hope soon it be pull

**Comment by onedolphin** [comment]
Thank you. This is for Magento 2 right?

**Comment by cbtrident** [comment]
for php 8.x and works with latest magento

**Comment by onedolphin** [comment]
Awesome, will try, thanks!

**Comment by onedolphin** [comment]
I have been trying to use the new one but without success, it shows everytime NO FILE and I have used the same settings as old Magmi... any idea?

File: magmi/plugins/5b5/general/attributesetimport/multi_dim_array.php [hunk 6]
```diff
@@ -75,6 +75,7 @@ public function offsetGet($name)
      *
      * @see ArrayIterator::offsetSet()
      */
+#[\ReturnTypeWillChange]
```
Comment (tmotyl) [inline magmi/plugins/5b5/general/attributesetimport/multi_dim_array.php:78]: why is this needed? what kind of error does it fix?

**Comment by tmotyl** [comment]
took your change @cbtrident and added some fixes. merged to magento2 branch.
Thanks!

File: magmi/plugins/5b5/general/attributesetimport/multi_dim_array.php (hunk 6 above)
Comment (cbtrident) [reply magmi/plugins/5b5/general/attributesetimport/multi_dim_array.php:78]: sorry can not remember think i when i was working out how to fix i was putting in comment to help me
as i was saying the to my self this may change to fix the error

**Comment by cbtrident** [comment]
thanks

**Comment by hazzy** [comment]
Thanks from me too!

**Comment by onedolphin** [comment]
Sorry still not working for me, it shows NO FILE :(

**Comment by hazzy** [comment]
> Sorry still not working for me, it shows NO FILE :(

How about a screenshot? Hard to help on not much info

**Comment by onedolphin** [comment]
Sure, will try to attach 2 here. Link below. Thanks for the help.

https://ibb.co/4TdNqrK
https://ibb.co/nMWMwZf

**Comment by cbtrident** [comment]
is your filesystem path right
did you make sure permision are correct to the file system for magmi
and did you save your new profile after you setup as it best to copy and name to a name you like in my case i did to zfs then change the setting. i i think you  should not play with default

**Comment by hazzy** [comment]
@onedolphin the NO FILE error comes from magmi_progress.php when it can't find the log file "progress.txt" in the "state" folder.

**Comment by onedolphin** [comment]
Hi, thanks for the help but still don't find the problem. I copy same of the errors in the log. I have not modified the default profile, I created a new one. I'm uploading the file and extracting and permissions are the same as always, I always installed Magmi this way and worked but this version is doing something different.

Here a part of error in the log file. Below this one is an error in the interface. 

[23-Jan-2023 17:04:00 UTC] PHP Parse error:  syntax error, unexpected 'public' (T_PUBLIC) in /home/do/magmi/engines/magmi_productimportengine.php on line 1922
[23-Jan-2023 17:10:26 UTC] PHP Parse error:  syntax error, unexpected 'public' (T_PUBLIC) in /home/do/magmi/engines/magmi_productimportengine.php on line 1922
[23-Jan-2023 17:12:15 UTC] PHP Deprecated:  Optional parameter $dorewrite declared before required parameter $store is implicitly treated as a required parameter in /home/do/magmi/plugins/extra/itemprocessors/itemindexer/otfindexer.php on line 232
[23-Jan-2023 17:12:15 UTC] PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /home/do/magmi/plugins/base/general/reindex/options_panel.php on line 21
[23-Jan-2023 17:12:15 UTC] PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /home/do/magmi/plugins/inc/magmi_plugin.php on line 146
[23-Jan-2023 17:12:15 UTC] PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /home/do/magmi/web/magmi_web_utils.php on line 6
[23-Jan-2023 17:12:15 UTC] PHP Deprecated:  addslashes(): Passing null to parameter #1 ($string) of type string is deprecated in /home/do/magmi/web/magmi_web_utils.php on line 9
[23-Jan-2023 17:12:15 UTC] PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /home/do/magmi/plugins/inc/magmi_plugin.php on line 146
[23-Jan-2023 17:12:15 UTC] PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /home/do/magmi/web/magmi_web_utils.php on line 6
[23-Jan-2023 17:12:15 UTC] PHP Deprecated:  addslashes(): Passing null to parameter #1 ($string) of type string is deprecated in /home/do/magmi/web/magmi_web_utils.php on line 9
[23-Jan-2023 17:12:15 UTC] PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /home/do/magmi/plugins/inc/magmi_plugin.php on line 146
[23-Jan-2023 17:12:15 UTC] PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /home/do/magmi/web/magmi_web_utils.php on line 6
[23-Jan-2023 17:12:15 UTC] PHP Deprecated:  addslashes(): Passing null to parameter #1 ($string) of type string is deprecated in /home/do/magmi/web/magmi_web_utils.php on line 9
[23-Jan-2023 17:12:15 UTC] PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /home/do/magmi/plugins/inc/magmi_plugin.php on line 146
[23-Jan-2023 17:12:15 UTC] PHP Deprecated:  explode(): Passing null to parameter #2 ($string) of type string is deprecated in /home/do/magmi/web/magmi_web_utils.php on line 6
[23-Jan-2023 17:12:15 UTC] PHP Deprecated:  addslashes(): Passing null to parameter #1 ($string) of type string is deprecated in /home/do/magmi/web/magmi_web_utils.php on line 9
[23-Jan-2023 17:13:00 UTC] PHP Deprecated:  Function strftime() is deprecated in /home/do/magmi/web/ajax_gettime.php on line 3
[23-Jan-2023 17:13:00 UTC] PHP Parse error:  syntax error, unexpected token "public" in /home/do/magmi/engines/magmi_productimportengine.php on line 1922

Error in interface:

Deprecated: Optional parameter $dorewrite declared before required parameter $store is implicitly treated as a required parameter in /home/do/magmi/plugins/extra/itemprocessors/itemindexer/otfindexer.php on line 232

**Comment by cbtrident** [comment]
Please test the main code now as it in main code now and what php you using

**Comment by onedolphin** [comment]
Sorry what do you mean main code? I set php 8.1 and then set again 7.4

**Comment by tmotyl** [comment]
@onedolphin please use branch "magento2" which is a main/default branch of this repository.
Clone the repository and checkout to magento2 branch (or git pull, if you're already on this branch).

**Comment by onedolphin** [comment]
@tmotyl Sorry I'm a newbie here. The site says the last update was 3 days ago. This is the one that will fix it? https://github.com/macopedia/magmi-m2/archive/refs/heads/magento2.zip

**Comment by tmotyl** [comment]
Yes, you can download zip and unpack it, but its not the best way to install plugins in Magento world.
Anyway, check if you have changes from this commit in place: https://github.com/macopedia/magmi-m2/commit/aa9f2c38ea4c132c828c162a58850b3bf21cc690

**Comment by onedolphin** [comment]
Sorry @tmotyl when I download the zip file is not the most updated files? Now when I downloaded a new one I cannot access magmi at all, I modified as usual the magmi.ini but cannot access, still shows I need to create a new magmi.ini file. Sorry new here wat github, still trying to see how this works. What do you mean to not install plugins in Magento, I'm only installing Magmi...

**Comment by cbtrident** [comment]
what os you install are you using apache or nginx
as there diffrent way to setup
the more information will help us to help you

**Comment by onedolphin** [comment]
@cbtrident I'm using LiteSpeed. I will be happy to provide any other details.

**Comment by cbtrident** [comment]
@onedolphin what os is that running on please
i will setup a system to test but never seen LiteSpeed

**Comment by onedolphin** [comment]
@cbtrident The OS is AlmaLinux 8.7.

**Comment by onedolphin** [comment]
Any ideas?

**Comment by cbtrident** [comment]
sorry waiting for my test key for LiteSpeed.

can you check what php you using pleasse doing php -v
i have to say i hate this linux i know it based on centos but not fun for me

**Comment by onedolphin** [comment]
Hi @cbtrident, please see what I got:

PHP 8.1.14 (cli) (built: Jan 11 2023 17:12:45) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.14, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.14, Copyright (c), by Zend Technologies

**Comment by onedolphin** [comment]
Help please...

**Comment by onedolphin** [comment]
> @onedolphin the NO FILE error comes from magmi_progress.php when it can't find the log file "progress.txt" in the "state" folder.

@hazzy Can I just create a progress.txt file? I did but now just shows plugins and startup below.
